knowledge-base/grid-select-all-onread.md

```diff
@@ -0,0 +1,147 @@
+---
+title: Select All Items With CheckBox when Using OnRead
+description: How Select All Items With CheckBox when Using OnRead in the Grid for Blazor
+type: how-to
+page_title: How Select All Items With CheckBox when Using OnRead
+slug: grid-kb-select-all-onread
+position: 
+tags: grid, selection, select all, checkbox, onread
+ticketid: 1562945
+res_type: kb
+---
+
+## Environment
+
+<table>
+    <tbody>
+        <tr>
+            <td>Product</td>
+            <td>Grid for Blazor</td>
+        </tr>
+    </tbody>
+</table>
+
+
+## Description
+
+I am binding the Grid through the `OnRead` event and I am using CheckBox selection with `SelectAllMode` set to `GridSelectAllMode.All`. However, when the user clicks the SelectAll CheckBox, only the items on the current page are selected, not all items in the dataset. How to ensure all items will be selected when clicking the SelectAll CheckBox.
+
+## Solution
+
+The described behavior is expected as when using the `OnRead` event, the Grid operates only with the current set/page of data. The component does not have information for all the items in your datasource and [thus it cannot select them all](slug:grid-selection-row#selection-and-paging).
+
+To ensure all item will be selected upon clicking the SelectAll CheckBox when using the `OnRead` event, you can implement a custom approach: 
+
+1. Use [`HeaderTemplate` for the `CheckboxColumn`](slug:components/grid/columns/checkbox#header-template) and add a custom [CheckBox component](slug:checkbox-overview) so you can have full control over its behavior.
+1. Handle the [`OnChange` event](slug:checkbox-events#onchange) of the CheckBox to track when the user checks/unchecks it to manage the selected items. 
+    - When the CheckBox is checked, request all the data from your datasource and assign it to the `SelectedItems` collection, so you have all items selected. You may want to cache all the data for further usage, so you don't need to request it additionally upon clicking the custom SelectAll CheckBox.
```

Comment:
This may not be acceptable if the data item count is too large. Another option is to raise a flag, so that all new items received in `OnRead` (e.g. during paging) are added to `SelectedItems`.